intra/doh/doh_test.go

```diff
@@ -406,6 +405,50 @@ func TestSendFailed(t *testing.T) {
 	}
 }
 
+// Test if DoH resolver IPs are confirmed and disconfirmed
+// when queries suceeded and fail, respectively.
+func TestDohIPConfirmDisconfirm(t *testing.T) {
+	u, _ := url.Parse(testURL)
+	doh, _ := NewTransport(testURL, ips, nil, nil, nil)
+	transport := doh.(*transport)
+	hostname := u.Hostname()
+	ipmap := transport.ips.Get(hostname)
+
+	// send a valid request to first have confirmed-ip set
+	res, _ := doh.Query(simpleQueryBytes)
+	mustUnpack(res)
+	ip1 := ipmap.Confirmed()
+
+	if ip1 == nil {
+		t.Errorf("IP not confirmed despite valid query to %s", u)
+	}
+
+	// simulate http-fail with doh server-ip set to previously confirmed-ip
+	rt := makeTestRoundTripper()
+	transport.client.Transport = rt
+	go func() {
+		req := <-rt.req
+		trace := httptrace.ContextClientTrace(req.Context())
+		trace.GotConn(httptrace.GotConnInfo{
+			Conn: &fakeConn{
+				remoteAddr: &net.TCPAddr{
+					IP:   ip1, // confirmed-ip from before
+					Port: 443,
+				}}})
+		rt.resp <- &http.Response{
+			StatusCode: 509, // some non-2xx status
+			Body:       nil,
+			Request:    &http.Request{URL: u},
+		}
+	}()
```

Comment:
`rt.err` is returned by `testRoundTripper.RoundTrip()`, which is used by `http.Client.Do()`, which is called [here](https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/80/files#diff-355c4ef8052beef46dc570a6de847f873c25583bc92b38fe9f382a9e398b255aR396).  I think this might not work for your test because the `testRoundTripper` ignores the `httptrace.ClientTrace`, which is used to collect the server identity, which is needed in order to disconfirm it.  So when the `testRoundTripper` is used, it may be that nothing is disconfirmed on failure (the failure is "too early").

Reply:
Right, such an intricate setup... Thanks!

To be clear: The current test case works like I expect it to. Without the current changeset it fails. With the current changeset it passes.

Will post output from `go test` in a bit when I'm at my PC.

Reply:
**with** the changeset:

```
➜  outline-go-tun2socks git:(ipmap) go test ./intra/doh  -v                       
=== RUN   TestSign
--- PASS: TestSign (0.01s)
=== RUN   TestSignNoIntermediate
--- PASS: TestSignNoIntermediate (0.00s)
=== RUN   TestNoAuth
--- PASS: TestNoAuth (0.00s)

... snip

=== RUN   TestDohIPConfirmDisconfirm
--- PASS: TestDohIPConfirmDisconfirm (0.09s)
=== RUN   TestListener
--- PASS: TestListener (0.00s)
=== RUN   TestAccept
--- PASS: TestAccept (0.00s)
=== RUN   TestAcceptFail
--- PASS: TestAcceptFail (0.00s)

... snip

=== RUN   TestAddEdnsPaddingCompressedOptQuery
--- PASS: TestAddEdnsPaddingCompressedOptQuery (0.00s)
=== RUN   TestAddEdnsPaddingCompressedPaddedQuery
--- PASS: TestAddEdnsPaddingCompressedPaddedQuery (0.00s)
=== RUN   TestServfail
--- PASS: TestServfail (0.00s)
PASS
ok  	github.com/Jigsaw-Code/outline-go-tun2socks/intra/doh	(cached)
```

**without**:

```
➜  outline-go-tun2socks git:(ipmap) ✗ go test ./intra/doh  -v       
=== RUN   TestSign
--- PASS: TestSign (0.01s)
=== RUN   TestSignNoIntermediate
--- PASS: TestSignNoIntermediate (0.00s)
=== RUN   TestNoAuth
--- PASS: TestNoAuth (0.00s)

... snip

=== RUN   TestDohIPConfirmDisconfirm
    doh_test.go:451: IP confirmed (8.8.8.8) despite err
--- FAIL: TestDohIPConfirmDisconfirm (0.08s)
=== RUN   TestListener
--- PASS: TestListener (0.00s)
=== RUN   TestAccept
--- PASS: TestAccept (0.00s)
=== RUN   TestAcceptFail
--- PASS: TestAcceptFail (0.00s)

... snip

=== RUN   TestAddEdnsPaddingCompressedOptQuery
--- PASS: TestAddEdnsPaddingCompressedOptQuery (0.00s)
=== RUN   TestAddEdnsPaddingCompressedPaddedQuery
--- PASS: TestAddEdnsPaddingCompressedPaddedQuery (0.00s)
=== RUN   TestServfail
--- PASS: TestServfail (0.00s)
FAIL
FAIL	github.com/Jigsaw-Code/outline-go-tun2socks/intra/doh	0.448s
FAIL
```